File: c++/BinaryTreeRightSideView.cpp
#include <iostream>
#include <vector>
#include "Helper.h" 
#include "TreeCreator.h" 

using namespace std;

class Solution {
	public:
		vector<int> rightSideView(TreeNode* root) {
			vector<int> output;

			if (root == NULL) {
				return output;
			}

			vector<TreeNode*> queue;
			queue.push_back(root);

			while(queue.size() > 0) {
				int curLen = queue.size();
				output.push_back(queue.back()->val);

				for (int i = 0; i < curLen; ++i) {
					TreeNode* curNode = queue[0];
					queue.erase(queue.begin());
					if (curNode->left)
						queue.push_back(curNode->left);
					if (curNode->right)
						queue.push_back(curNode->right);
				}
			}

			return output;
		}
};

int main() {
	Solution sol;
        TreeCreator creator;
        vector<int> nodeList = {7, 3, 15, null, null, 9, 20};
        TreeNode* root = creator.createTree(nodeList);

	vector<int> output = sol.rightSideView(root);
	Helper::print_vector(output);
	
	return 0;
}

File: python3/BinaryTreeRightSideView.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if (root == None):
            return []
        
        queue = [root]
        
        output = []
        while (len(queue) > 0):
            curLen = len(queue)
            output.append(queue[-1].val)
            for i in range(curLen):
                item = queue.pop(0)
                if item.left:
                    queue.append(item.left)
                if item.right:
                    queue.append(item.right)
        
        return output
